db/src/cache.rs

```diff
@@ -861,16 +896,227 @@ impl AttributeCaches {
+                    interpose!(item, fts,
+                               { qb.push_sql(&item.to_string()) },
+                               { qb.push_sql(", ") });
+                    qb.push_sql(")");
+
+                    self.forward_cached_attributes.extend(fts.iter());
+                }
+                qb.push_sql(" ORDER BY a ASC, e ASC");
+            },
+        };
+
+        let SQLQuery { sql, args } = qb.finish();
+        assert!(args.is_empty());                       // TODO: we know there are never args, but we'd like to run this query 'properly'.
+        let mut stmt = sqlite.prepare(sql.as_str())?;
+        let replacing = false;
+        self.repopulate_from_aevt(schema, &mut stmt, vec![], replacing)
+    }
+
+    pub fn forward_attribute_cache_for_attribute<'a, 's>(&'a self, schema: &'s Schema, a: Entid) -> Option<&'a AttributeCache> {
+        if !self.forward_cached_attributes.contains(&a) {
+            return None;
+        }
+        schema.attribute_for_entid(a)
+              .and_then(|attr|
+                if attr.multival {
+                    self.multi_vals.get(&a).map(|v| v as &AttributeCache)
+                } else {
+                    self.single_vals.get(&a).map(|v| v as &AttributeCache)
+                })
+    }
+
+    // Ensure that `entities` is unique.
+    pub fn extend_cache_for_entities_and_attributes<'s, 'c>(&mut self,
+                                                            schema: &'s Schema,
+                                                            sqlite: &'c rusqlite::Connection,
+                                                            mut attrs: AttributeSpec,
+                                                            entities: &Vec<Entid>) -> Result<()> {
+        // TODO: Exclude any entities for which every attribute is known.
+        // TODO: initialize from an existing (complete) AttributeCache.
+
+        // Exclude any attributes for which every entity's value is already known.
+        match &mut attrs {
+            &mut AttributeSpec::All => {
+                // If we're caching all attributes, there's nothing we can exclude.
+            },
+            &mut AttributeSpec::Specified { ref mut non_fts, ref mut fts } => {
+                // Remove any attributes for which all entities are present in the cache (even
+                // as a 'miss').
+                let exclude_missing = |vec: &mut Vec<Entid>| {
+                    vec.retain(|a| {
+                        if let Some(attr) = schema.attribute_for_entid(*a) {
+                            if !self.forward_cached_attributes.contains(a) {
+                                // The attribute isn't cached at all. Do the work for all entities.
+                                return true;
+                            }
+
+                            // Return true if there are any entities missing for this attribute.
+                            if attr.multival {
```

Comment:
Either Rust won't let you do it at all, or you need to reduce it to an `AttributeCache` trait object, yes. The types of each line are different, because one works with single-value caches and the other with multi-value caches.